beacon-chain/validator-monitor/service.go

```diff
@@ -0,0 +1,359 @@
+		},
+		[]string{
+			"validator_index",
+		},
+	)
+	// aggregationCounter used to track aggregations
+	aggregationCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "monitor",
+			Name:      "aggregations",
+			Help:      "Number of aggregation duties performed",
+		},
+		[]string{
+			"validator_index",
+		},
+	)
+	// syncCommitteeContributionCounter used to track sync committee
+	// contributions
+	syncCommitteeContributionCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "monitor",
+			Name:      "sync_committee_contributions",
+			Help:      "Number of Sync committee contributions performed",
+		},
+		[]string{
+			"validator_index",
+		},
+	)
+	// Error when event feed data is not statefeed.SyncedData
+	errorNotSyncedData = errors.New("Event feed data is not of type *statefeed.SyncedData")
+
+	// Error when the context is closed while waiting for sync
+	errorContextClosedWhileWaiting = errors.New("Context closed while waiting for beacon to sync to latest Head")
+)
+
+// ValidatorLatestPerformance keeps track of the latest participation of the validator
+type ValidatorLatestPerformance struct {
+	attestedSlot  types.Slot
+	inclusionSlot types.Slot
+	timelySource  bool
+	timelyTarget  bool
+	timelyHead    bool
+	balance       uint64
+	balanceChange uint64
+}
+
+// ValidatorAggregatedPerformance keeps track of the accumulated performance of
+// the validator since launch
+type ValidatorAggregatedPerformance struct {
+	startEpoch                     types.Epoch
+	startBalance                   uint64
+	totalAttestedCount             uint64
+	totalRequestedCount            uint64
+	totalDistance                  uint64
+	totalCorrectSource             uint64
+	totalCorrectTarget             uint64
+	totalCorrectHead               uint64
+	totalProposedCount             uint64
+	totalAggregations              uint64
+	totalSyncComitteeContributions uint64
+	totalSyncComitteeAggregations  uint64
+}
+
+// ValidatorMonitorConfig contains the list of validator indices that the
+// validator-monitor service tracks, as well as the event feed notifier that the
+// monitor needs to subscribe.
+type ValidatorMonitorConfig struct {
+	TrackedValidators   []types.ValidatorIndex
+	StateNotifier       statefeed.Notifier
+	AttestationNotifier operation.Notifier
+	StateGen            stategen.StateManager
+	HeadFetcher         blockchain.HeadFetcher
+}
+
+// Service is the main structure that tracks validators and reports logs and
+// metrics of their performances throughout their lifetime.
+type Service struct {
+	config                      *ValidatorMonitorConfig
+	ctx                         context.Context
+	cancel                      context.CancelFunc
+	latestPerformance           map[types.ValidatorIndex]ValidatorLatestPerformance
+	aggregatedPerformance       map[types.ValidatorIndex]ValidatorAggregatedPerformance
+	trackedSyncCommitteeIndices map[types.ValidatorIndex][]types.CommitteeIndex
```

Comment:
i think cache pkg's `SyncCommitteeCache` does exactly this, but maybe it's easier to double track here

Reply:
Oh but here I want the list of indices that I track that are in the sync committee, otherwise I'll need to take an intersection on every slot. Anyway the call is cheap cause I'm taking the committee from the cache, and only once per period.